src/Common/src/Interop/OSX/Interop.libproc.cs

```diff
@@ -0,0 +1,542 @@
+            {
+                throw new ArgumentOutOfRangeException("pid");
+            }
+
+            // Query for an estimation about the size of the buffer we will need. This seems
+            // to add some padding from the real number, so we don't need to do that
+            int result = proc_pidinfo(pid, PROC_PIDLISTFDS, 0, (proc_fdinfo*)null, 0);
+            if (result <= 0)
+            {
+                throw new System.ComponentModel.Win32Exception();
+            }
+
+            proc_fdinfo[] fds = null;
+            int size = (int)(result / Marshal.SizeOf<proc_fdinfo>());
+
+            // Just in case the app opened a ton of handles between when we asked and now,
+            // make sure we retry if our buffer is filled
+            do
+            {
+                fds = new proc_fdinfo[size];
+                fixed (proc_fdinfo* pFds = fds)
+                {
+                    result = proc_pidinfo(pid, PROC_PIDLISTFDS, 0, pFds, Marshal.SizeOf<proc_fdinfo>() * fds.Length);
+                }
+
+                if (result <= 0)
+                {
+                    throw new System.ComponentModel.Win32Exception(Marshal.GetLastWin32Error());
+                }
+                else
+                {
+                    size *= 2;
+                }
+            }
+            while (result == (fds.Length * Marshal.SizeOf<proc_fdinfo>()));
+
+            Debug.Assert((result % Marshal.SizeOf<proc_fdinfo>()) == 0);
+
+            return fds;
+        }
+
+        /// <summary>
+        /// Gets the full path to the executable file identified by the specified PID
+        /// </summary>
+        /// <param name="pid">The PID of the running process</param>
+        /// <param name="buffer">A pointer to an allocated block of memory that will be filled with the process path</param>
+        /// <param name="bufferSize">The size of the buffer, should be PROC_PIDPATHINFO_MAXSIZE</param>
+        /// <returns>Returns the length of the path returned on success</returns>
+        [DllImport(Interop.Libraries.libproc, CharSet = CharSet.Ansi, SetLastError = true)]
+        private static unsafe extern int proc_pidpath(
+            int pid, 
+            byte* buffer, 
+            uint bufferSize);
+
+        /// <summary>
+        /// Gets the full path to the executable file identified by the specified PID
+        /// </summary>
+        /// <param name="pid">The PID of the running process</param>
+        /// <returns>Returns the full path to the process executable</returns>
+        internal static unsafe string proc_pidpath(int pid)
+        {
+            // Negative PIDs are invalid
+            if (pid < 0)
+            {
+                throw new ArgumentOutOfRangeException("pid", SR.NegativePidNotSupported);
+            }
+
+            // The path is a fixed buffer size, so use that and trim it after
+            int result = 0;
+            byte[] buffer = new byte[PROC_PIDPATHINFO_MAXSIZE];
```

Comment:
Consider using a stack-allocated buffer here, e.g.
```C#
byte* buffer = stackalloc byte[PROC_PIDPATHINFO_MAXSIZE];
result = proc_pidpath(pid, pBuffer, (uint)PROC_PIDPATHINFO_MAXSIZE);
if (result <= 0)
{
    throw ...;
}
return Encoding.UTF8.GetString(buffer, result);
```
(This assumes that result is the number of bytes filled in; you could adjust accordingly if it's not.)